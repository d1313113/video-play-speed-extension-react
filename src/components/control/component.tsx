import React, { FC } from "react";
import { Slider, Button } from "antd";
import {
  PlayCircleOutlined,
  PauseCircleOutlined,
  AudioMutedOutlined,
  AudioOutlined
} from "@ant-design/icons";
import { VideoStatus } from "@/content-script/actions/videosPlaybackRate";
import "./styles.scss";

const RATES = [1, 2, 3, 4, 5];

export interface ControlProps {
  videoList: VideoStatus[];
  onSetRate: (idx: number, value: number) => Promise<void>;
  onPlay: (idx: number) => Promise<void>;
  onPause: (idx: number) => Promise<void>;
  onMuted: (idx: number, muted: boolean) => Promise<void>;
}

export const Control: FC<ControlProps> = (props: ControlProps) => {
  const handleSetVideoRate = (index: number, value: number): void => {
    props.onSetRate(index, value);
  };

  const handleClickRate = (index: number, rate: number): void => {
    props.onSetRate(index, rate);
  };

  const { videoList } = props;

  return (
    <main className="control">
      <div className="title">可拖动step调整速度</div>
      <section className="control-main">
        {videoList.map((item, index) => (
          <div className="video-item" key={index}>
            <Slider
              onChange={(value: number): void =>
                handleSetVideoRate(index, value)
              }
              min={0.1}
              max={5}
              step={0.1}
              defaultValue={1}
              value={item.playbackRate}
            />
            <div className="rate-wrapper">
              {item.paused ? (
                <PlayCircleOutlined
                  className="icon-control"
                  onClick={(): Promise<void> => props.onPlay(index)}
                />
              ) : (
                <PauseCircleOutlined
                  className="icon-control"
                  onClick={(): Promise<void> => props.onPause(index)}
                />
              )}
              {item.muted ? (
                <AudioMutedOutlined
                  className="icon-control middle"
                  onClick={(): Promise<void> =>
                    props.onMuted(index, !item.muted)
                  }
                />
              ) : (
                <AudioOutlined
                  className="icon-control middle"
                  onClick={(): Promise<void> =>
                    props.onMuted(index, !item.muted)
                  }
                />
              )}
              {RATES.map(rate => (
                <Button
                  key={rate}
                  onClick={(): void => handleClickRate(index, rate)}
                  size="small"
                  type={rate === 1 ? "primary" : "default"}
                  disabled={rate === item.playbackRate}
                >
                  {rate === 1 ? "Reset" : `✖️${rate}`}
                </Button>
              ))}
            </div>
          </div>
        ))}
      </section>
    </main>
  );
};
