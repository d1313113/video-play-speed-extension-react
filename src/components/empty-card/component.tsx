import React, { FC } from "react";
import { Empty } from "antd";
import "./styles.scss";

export const EmptyCard: FC = () => (
  <section className="empty-wrapper">
    <Empty className="empty" description="当前页面未找到视频:)" />
  </section>
);
