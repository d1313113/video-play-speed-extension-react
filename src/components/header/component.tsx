import React, { FC, useState } from "react";
import { ReloadOutlined } from "@ant-design/icons";
import "./styles.scss";

interface HeaderProps {
  onReload: () => void;
}

export const Header: FC<HeaderProps> = (props: HeaderProps) => {
  const [spin, setSpin] = useState<boolean>(false);

  const handleReload = (): void => {
    setSpin(!spin);
    if (!spin) {
      props.onReload();
      setTimeout((): void => {
        setSpin(false);
      }, 1000);
    }
  };

  return (
    <header className="header-wrapper">
      <div className="header-content">
        <div className="title">视频播放速度调节插件</div>
        <div className="sub-title">可以让你轻松调整视频播放速度</div>
      </div>
      <ReloadOutlined
        onClick={handleReload}
        className="icon-reload"
        spin={spin}
        style={{
          fontSize: "20px"
        }}
      />
    </header>
  );
};
