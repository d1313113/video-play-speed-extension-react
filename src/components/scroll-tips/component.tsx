import React, { FC } from "react";
import { DoubleRightOutlined } from "@ant-design/icons";
import "./styles.scss";

export const ScrollTips: FC = () => (
  <section className="tips-wrapper">
    <div className="tips">Scroll to load more</div>
    <DoubleRightOutlined
      className="icon-tips"
      rotate={90}
      style={{ fontSize: "20px" }}
    />
  </section>
);
