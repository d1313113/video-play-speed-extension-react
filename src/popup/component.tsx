import React, { FC, useEffect, useState } from "react";
import { browser } from "webextension-polyfill-ts";
import { throttle } from "lodash";
import getCurrentTab from "@/utils/getCurrentTab";
import {
  actionTypes,
  VideoStatus,
  $$
} from "@/content-script/actions/videosPlaybackRate";

import { Header } from "@/components/header";
import { EmptyCard } from "@/components/empty-card";
import { Control } from "@/components/control";
import { ScrollTips } from "@/components/scroll-tips";
import "./styles.scss";

export const Popup: FC = () => {
  const [videoList, setVideoList] = useState<VideoStatus[]>([]);
  const [scrollTop, setScrollTop] = useState<number>(0);

  const initPopUp = async (): Promise<void> => {
    console.log(11111);
    const { id: tabId } = await getCurrentTab();
    if (tabId != null) {
      const resp: VideoStatus[] = await browser.tabs.sendMessage(tabId, {
        action: actionTypes.GET_VIDEOS_STATUS
      });
      setVideoList(resp);
    }
  };

  const handleSetVideoRate = async (
    idx: number,
    value: number
  ): Promise<void> => {
    const { id: tabId } = await getCurrentTab();

    if (tabId != null) {
      await browser.tabs.sendMessage(tabId, {
        action: actionTypes.SET_VIDEO_PLAYBACK_RATE,
        idx,
        value
      });
      initPopUp();
    }
  };

  const handleSetVideoPlay = async (idx: number): Promise<void> => {
    const { id: tabId } = await getCurrentTab();

    if (tabId != null) {
      await browser.tabs.sendMessage(tabId, {
        action: actionTypes.SET_VIDEO_PLAY,
        idx
      });
      initPopUp();
    }
  };

  const handleSetVideoPause = async (idx: number): Promise<void> => {
    const { id: tabId } = await getCurrentTab();

    if (tabId != null) {
      await browser.tabs.sendMessage(tabId, {
        action: actionTypes.SET_VIDEO_PAUSE,
        idx
      });
      initPopUp();
    }
  };

  const handleScroll = (html: HTMLElement): void => {
    try {
      const { scrollTop } = html;
      setScrollTop(scrollTop);
    } catch (e) {
      console.error(e);
    }
  };

  // Sends the `popupMounted` event
  useEffect(() => {
    browser.runtime.sendMessage({ popupMounted: true });
    initPopUp();
    const html = $$<HTMLElement>("html")[0];
    const _throttle = throttle((): void => handleScroll(html), 100);
    window.addEventListener("scroll", _throttle, false);

    return (): void => {
      window.removeEventListener("scroll", _throttle, false);
    };
  }, []);

  // Renders the component tree
  return (
    <div className="popup-container">
      <div className="container">
        <Header onReload={initPopUp} />
        {!videoList.length ? (
          <EmptyCard />
        ) : (
          <Control
            videoList={videoList}
            onSetRate={handleSetVideoRate}
            onPlay={handleSetVideoPlay}
            onPause={handleSetVideoPause}
          />
        )}
        {!!videoList.length && videoList.length > 5 && scrollTop < 30 && (
          <ScrollTips />
        )}
      </div>
    </div>
  );
};
