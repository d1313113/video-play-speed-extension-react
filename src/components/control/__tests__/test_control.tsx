import * as React from "react";
import { Control, ControlProps } from "../component";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";

it("component renders", () => {
  const videoList = [1, 2, 3].map(item => ({
    playbackRate: item,
    paused: false,
    muted: false
  }));
  const onSetRate: jest.Mock<
    Promise<void>,
    [number, number]
  > = jest.fn((idx: number, value: number) => Promise.resolve());
  const onPlay: jest.Mock<Promise<void>, [number]> = jest.fn((idx: number) =>
    Promise.resolve()
  );
  const onPause: jest.Mock<Promise<void>, [number]> = jest.fn((idx: number) =>
    Promise.resolve()
  );
  const props: ControlProps = {
    videoList,
    onSetRate,
    onPlay,
    onPause
  };

  const tree:
    | ReactTestRendererJSON
    | ReactTestRendererJSON[]
    | null = renderer.create(<Control {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});
