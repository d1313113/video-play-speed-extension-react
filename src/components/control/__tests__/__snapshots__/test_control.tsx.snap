// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders 1`] = `
<main
  className="control"
>
  <div
    className="title"
  >
    可拖动step调整速度
  </div>
  <section
    className="control-main"
  >
    <div
      className="video-item"
    >
      <div
        className="ant-slider"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="ant-slider-rail"
          style={Object {}}
        />
        <div
          className="ant-slider-track"
          style={
            Object {
              "left": "0%",
              "right": "auto",
              "width": "18.36734693877551%",
            }
          }
        />
        <div
          className="ant-slider-step"
        />
        <div
          aria-disabled={false}
          aria-valuemax={5}
          aria-valuemin={0.1}
          aria-valuenow={1}
          className="ant-slider-handle"
          onBlur={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="slider"
          style={
            Object {
              "left": "18.36734693877551%",
              "right": "auto",
              "transform": "translateX(-50%)",
            }
          }
          tabIndex={0}
        />
        <div
          className="ant-slider-mark"
        />
      </div>
      <div
        className="rate-wrapper"
      >
        <span
          aria-label="pause-circle"
          className="anticon anticon-pause-circle icon-control"
          onClick={[Function]}
          role="img"
          tabIndex={-1}
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="pause-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372zm-88-532h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8zm224 0h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8z"
            />
          </svg>
        </span>
        <span
          aria-label="audio"
          className="anticon anticon-audio icon-control middle"
          onClick={[Function]}
          role="img"
          tabIndex={-1}
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="audio"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M842 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8 0 140.3-113.7 254-254 254S258 594.3 258 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8 0 168.7 126.6 307.9 290 327.6V884H326.7c-13.7 0-24.7 14.3-24.7 32v36c0 4.4 2.8 8 6.2 8h407.6c3.4 0 6.2-3.6 6.2-8v-36c0-17.7-11-32-24.7-32H548V782.1c165.3-18 294-158 294-328.1zM512 624c93.9 0 170-75.2 170-168V232c0-92.8-76.1-168-170-168s-170 75.2-170 168v224c0 92.8 76.1 168 170 168zm-94-392c0-50.6 41.9-92 94-92s94 41.4 94 92v224c0 50.6-41.9 92-94 92s-94-41.4-94-92V232z"
            />
          </svg>
        </span>
        <button
          className="ant-btn ant-btn-primary ant-btn-sm"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          <span>
            Reset
          </span>
        </button>
        <button
          className="ant-btn ant-btn-default ant-btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            ✖️2
          </span>
        </button>
        <button
          className="ant-btn ant-btn-default ant-btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            ✖️3
          </span>
        </button>
        <button
          className="ant-btn ant-btn-default ant-btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            ✖️4
          </span>
        </button>
        <button
          className="ant-btn ant-btn-default ant-btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            ✖️5
          </span>
        </button>
      </div>
    </div>
    <div
      className="video-item"
    >
      <div
        className="ant-slider"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="ant-slider-rail"
          style={Object {}}
        />
        <div
          className="ant-slider-track"
          style={
            Object {
              "left": "0%",
              "right": "auto",
              "width": "38.77551020408163%",
            }
          }
        />
        <div
          className="ant-slider-step"
        />
        <div
          aria-disabled={false}
          aria-valuemax={5}
          aria-valuemin={0.1}
          aria-valuenow={2}
          className="ant-slider-handle"
          onBlur={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="slider"
          style={
            Object {
              "left": "38.77551020408163%",
              "right": "auto",
              "transform": "translateX(-50%)",
            }
          }
          tabIndex={0}
        />
        <div
          className="ant-slider-mark"
        />
      </div>
      <div
        className="rate-wrapper"
      >
        <span
          aria-label="pause-circle"
          className="anticon anticon-pause-circle icon-control"
          onClick={[Function]}
          role="img"
          tabIndex={-1}
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="pause-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372zm-88-532h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8zm224 0h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8z"
            />
          </svg>
        </span>
        <span
          aria-label="audio"
          className="anticon anticon-audio icon-control middle"
          onClick={[Function]}
          role="img"
          tabIndex={-1}
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="audio"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M842 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8 0 140.3-113.7 254-254 254S258 594.3 258 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8 0 168.7 126.6 307.9 290 327.6V884H326.7c-13.7 0-24.7 14.3-24.7 32v36c0 4.4 2.8 8 6.2 8h407.6c3.4 0 6.2-3.6 6.2-8v-36c0-17.7-11-32-24.7-32H548V782.1c165.3-18 294-158 294-328.1zM512 624c93.9 0 170-75.2 170-168V232c0-92.8-76.1-168-170-168s-170 75.2-170 168v224c0 92.8 76.1 168 170 168zm-94-392c0-50.6 41.9-92 94-92s94 41.4 94 92v224c0 50.6-41.9 92-94 92s-94-41.4-94-92V232z"
            />
          </svg>
        </span>
        <button
          className="ant-btn ant-btn-primary ant-btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            Reset
          </span>
        </button>
        <button
          className="ant-btn ant-btn-default ant-btn-sm"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          <span>
            ✖️2
          </span>
        </button>
        <button
          className="ant-btn ant-btn-default ant-btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            ✖️3
          </span>
        </button>
        <button
          className="ant-btn ant-btn-default ant-btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            ✖️4
          </span>
        </button>
        <button
          className="ant-btn ant-btn-default ant-btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            ✖️5
          </span>
        </button>
      </div>
    </div>
    <div
      className="video-item"
    >
      <div
        className="ant-slider"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="ant-slider-rail"
          style={Object {}}
        />
        <div
          className="ant-slider-track"
          style={
            Object {
              "left": "0%",
              "right": "auto",
              "width": "59.183673469387756%",
            }
          }
        />
        <div
          className="ant-slider-step"
        />
        <div
          aria-disabled={false}
          aria-valuemax={5}
          aria-valuemin={0.1}
          aria-valuenow={3}
          className="ant-slider-handle"
          onBlur={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="slider"
          style={
            Object {
              "left": "59.183673469387756%",
              "right": "auto",
              "transform": "translateX(-50%)",
            }
          }
          tabIndex={0}
        />
        <div
          className="ant-slider-mark"
        />
      </div>
      <div
        className="rate-wrapper"
      >
        <span
          aria-label="pause-circle"
          className="anticon anticon-pause-circle icon-control"
          onClick={[Function]}
          role="img"
          tabIndex={-1}
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="pause-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372zm-88-532h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8zm224 0h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8z"
            />
          </svg>
        </span>
        <span
          aria-label="audio"
          className="anticon anticon-audio icon-control middle"
          onClick={[Function]}
          role="img"
          tabIndex={-1}
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="audio"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M842 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8 0 140.3-113.7 254-254 254S258 594.3 258 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8 0 168.7 126.6 307.9 290 327.6V884H326.7c-13.7 0-24.7 14.3-24.7 32v36c0 4.4 2.8 8 6.2 8h407.6c3.4 0 6.2-3.6 6.2-8v-36c0-17.7-11-32-24.7-32H548V782.1c165.3-18 294-158 294-328.1zM512 624c93.9 0 170-75.2 170-168V232c0-92.8-76.1-168-170-168s-170 75.2-170 168v224c0 92.8 76.1 168 170 168zm-94-392c0-50.6 41.9-92 94-92s94 41.4 94 92v224c0 50.6-41.9 92-94 92s-94-41.4-94-92V232z"
            />
          </svg>
        </span>
        <button
          className="ant-btn ant-btn-primary ant-btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            Reset
          </span>
        </button>
        <button
          className="ant-btn ant-btn-default ant-btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            ✖️2
          </span>
        </button>
        <button
          className="ant-btn ant-btn-default ant-btn-sm"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          <span>
            ✖️3
          </span>
        </button>
        <button
          className="ant-btn ant-btn-default ant-btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            ✖️4
          </span>
        </button>
        <button
          className="ant-btn ant-btn-default ant-btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            ✖️5
          </span>
        </button>
      </div>
    </div>
  </section>
</main>
`;
